File: test/test_daily_report_metro_fix.py
"""
Test script to verify Daily Report Metro Municipality fix
Analyzes the existing Daily Report to check if Metro Municipalities are correctly handled
"""

import pandas as pd
import sys
from pathlib import Path

def test_daily_report_metro_fix():
    print('🧪 Testing Daily Report Metro Municipality Fix\n')
    print('=' * 80)

    try:
        # Load the existing report
        report_path = Path(__file__).parent.parent / 'solo' / 'Daily_Report_2025-11-18 (1).xlsx'
        print(f'📂 Loading report from: {report_path}\n')

        # Read the Municipality-District Analysis sheet
        df = pd.read_excel(report_path, sheet_name='Municipality-District Analysis')
        
        # Extract data
        municipalities = []
        metro_municipalities = []
        gauteng_municipalities = []
        current_province = ''
        province_data = {}

        print('📊 Analyzing sheet data...\n')

        for idx, row in df.iterrows():
            municipality_name = str(row.iloc[0]) if pd.notna(row.iloc[0]) else ''
            iec_wards = row.iloc[1] if pd.notna(row.iloc[1]) else None

            # Check if it's a province header (no data in other columns)
            if pd.isna(iec_wards) and 'Total' not in municipality_name and municipality_name:
                current_province = municipality_name
                province_data[current_province] = []
                print(f'📍 Province: {current_province}')
                continue

            # Skip totals rows
            if 'Total' in municipality_name:
                continue

            # Check for Metro Municipalities
            if 'Metropolitan Municipality' in municipality_name:
                metro_municipalities.append({
                    'name': municipality_name,
                    'province': current_province,
                    'wards': iec_wards
                })
                print(f'   ❌ FOUND METRO: {municipality_name} ({iec_wards} wards)')
            elif municipality_name:
                municipalities.append(municipality_name)
                if current_province:
                    province_data[current_province].append(municipality_name)
                if current_province == 'Gauteng':
                    gauteng_municipalities.append(municipality_name)
                    print(f'   ✅ {municipality_name}')

        # Results
        print('\n' + '=' * 80)
        print('📊 TEST RESULTS')
        print('=' * 80)

        print(f'\n✅ Total municipalities/sub-regions found: {len(municipalities)}')
        print(f'✅ Gauteng municipalities/sub-regions: {len(gauteng_municipalities)}')

        if len(metro_municipalities) == 0:
            print('\n✅ SUCCESS: No Metro Municipalities found in report')
            print('   Metro Municipalities are correctly excluded')
            print('   Only sub-regions are shown')
        else:
            print(f'\n❌ FAILURE: Found {len(metro_municipalities)} Metro Municipalities:')
            for metro in metro_municipalities:
                print(f'   - {metro["name"]} (Province: {metro["province"]}, Wards: {metro["wards"]})')
            print('\n   These should NOT appear in the report!')
            print('   Only their sub-regions should be shown.')

        print('\n📋 Gauteng Municipalities/Sub-Regions:')
        for muni in gauteng_municipalities:
            print(f'   - {muni}')

        print('\n📊 Province Summary:')
        for province, munis in province_data.items():
            if province:
                print(f'   {province}: {len(munis)} municipalities/sub-regions')

        print('\n' + '=' * 80)
        print('🎯 EXPECTED BEHAVIOR:')
        print('=' * 80)
        print('❌ Should NOT see:')
        print('   - City of Johannesburg Metropolitan Municipality')
        print('   - City of Tshwane Metropolitan Municipality')
        print('   - Ekurhuleni Metropolitan Municipality')
        print('\n✅ Should see:')
        print('   - Emfuleni Sub-Region')
        print('   - Lesedi Sub-Region')
        print('   - Merafong City Sub-Region')
        print('   - Midvaal Sub-Region')
        print('   - Mogale City Sub-Region')
        print('   - Rand West City Sub-Region')
        print('   - (and other Gauteng sub-regions)')

        print('\n' + '=' * 80)

        if len(metro_municipalities) == 0:
            print('✅ TEST PASSED: Metro Municipality fix is working correctly!')
            print('\n💡 The report correctly shows only sub-regions, not metro municipalities.')
            return True
        else:
            print('❌ TEST FAILED: Metro Municipalities still appearing in report')
            print('\n💡 The SQL query needs to be updated to exclude Metropolitan municipalities.')
            print('   Filter: WHERE municipality_type != \'Metropolitan\'')
            return False

    except Exception as error:
        print(f'\n❌ Error during test: {str(error)}')
        import traceback
        traceback.print_exc()
        return False

if __name__ == '__main__':
    success = test_daily_report_metro_fix()
    sys.exit(0 if success else 1)

File: test/test_daily_report_metro_fix.js
/**
 * Test script to verify Daily Report Metro Municipality fix
 * Analyzes the existing Daily Report to check if Metro Municipalities are correctly handled
 */

const ExcelJS = require('exceljs');
const path = require('path');

async function testDailyReportMetroFix() {
  console.log('🧪 Testing Daily Report Metro Municipality Fix\n');
  console.log('='.repeat(80));

  try {
    // Load the existing report
    const reportPath = path.join(__dirname, '..', 'solo', 'Daily_Report_2025-11-18 (1).xlsx');
    console.log(`📂 Loading report from: ${reportPath}\n`);

    const workbook = new ExcelJS.Workbook();
    await workbook.xlsx.readFile(reportPath);

    const sheet = workbook.getWorksheet('Municipality-District Analysis');
    if (!sheet) {
      throw new Error('Municipality-District Analysis sheet not found');
    }

    // Extract data
    const municipalities = [];
    let currentProvince = '';
    const metroMunicipalities = [];
    const gautengMunicipalities = [];
    const provinceData = {};

    sheet.eachRow((row, rowNumber) => {
      if (rowNumber === 1) return; // Skip header

      const municipalityName = row.getCell(1).value?.toString() || '';
      const iecWards = row.getCell(2).value;

      // Check if it's a province header (no data in other columns)
      if (!iecWards && !municipalityName.includes('Total')) {
        currentProvince = municipalityName;
        provinceData[currentProvince] = [];
        console.log(`\n📍 Province: ${currentProvince}`);
        return;
      }

      // Skip totals rows
      if (municipalityName.includes('Total')) {
        return;
      }

      // Check for Metro Municipalities
      if (municipalityName.includes('Metropolitan Municipality')) {
        metroMunicipalities.push({
          name: municipalityName,
          province: currentProvince,
          wards: iecWards
        });
        console.log(`   ❌ FOUND METRO: ${municipalityName} (${iecWards} wards)`);
      } else if (municipalityName) {
        municipalities.push(municipalityName);
        if (currentProvince) {
          provinceData[currentProvince].push(municipalityName);
        }
        if (currentProvince === 'Gauteng') {
          gautengMunicipalities.push(municipalityName);
          console.log(`   ✅ ${municipalityName}`);
        }
      }
    });

    // Results
    console.log('\n' + '='.repeat(80));
    console.log('📊 TEST RESULTS');
    console.log('='.repeat(80));

    console.log(`\n✅ Total municipalities/sub-regions found: ${municipalities.length}`);
    console.log(`✅ Gauteng municipalities/sub-regions: ${gautengMunicipalities.length}`);

    if (metroMunicipalities.length === 0) {
      console.log('\n✅ SUCCESS: No Metro Municipalities found in report');
      console.log('   Metro Municipalities are correctly excluded');
      console.log('   Only sub-regions are shown');
    } else {
      console.log(`\n❌ FAILURE: Found ${metroMunicipalities.length} Metro Municipalities:`);
      metroMunicipalities.forEach(metro => {
        console.log(`   - ${metro.name} (Province: ${metro.province}, Wards: ${metro.wards})`);
      });
      console.log('\n   These should NOT appear in the report!');
      console.log('   Only their sub-regions should be shown.');
    }

    console.log('\n📋 Gauteng Municipalities/Sub-Regions:');
    gautengMunicipalities.forEach(muni => {
      console.log(`   - ${muni}`);
    });

    console.log('\n📊 Province Summary:');
    Object.keys(provinceData).forEach(province => {
      if (province) {
        console.log(`   ${province}: ${provinceData[province].length} municipalities/sub-regions`);
      }
    });

    console.log('\n' + '='.repeat(80));
    console.log('🎯 EXPECTED BEHAVIOR:');
    console.log('='.repeat(80));
    console.log('❌ Should NOT see:');
    console.log('   - City of Johannesburg Metropolitan Municipality');
    console.log('   - City of Tshwane Metropolitan Municipality');
    console.log('   - Ekurhuleni Metropolitan Municipality');
    console.log('\n✅ Should see:');
    console.log('   - Emfuleni Sub-Region');
    console.log('   - Lesedi Sub-Region');
    console.log('   - Merafong City Sub-Region');
    console.log('   - Midvaal Sub-Region');
    console.log('   - Mogale City Sub-Region');
    console.log('   - Rand West City Sub-Region');
    console.log('   - (and other Gauteng sub-regions)');

    console.log('\n' + '='.repeat(80));

    if (metroMunicipalities.length === 0) {
      console.log('✅ TEST PASSED: Metro Municipality fix is working correctly!');
      console.log('\n💡 The report correctly shows only sub-regions, not metro municipalities.');
      return true;
    } else {
      console.log('❌ TEST FAILED: Metro Municipalities still appearing in report');
      console.log('\n💡 The SQL query needs to be updated to exclude Metropolitan municipalities.');
      console.log('   Filter: WHERE municipality_type != \'Metropolitan\'');
      return false;
    }

  } catch (error) {
    console.error('\n❌ Error during test:', error.message);
    console.error(error.stack);
    return false;
  }
}

// Run the test
testDailyReportMetroFix()
  .then(success => {
    process.exit(success ? 0 : 1);
  })
  .catch(error => {
    console.error('Fatal error:', error);
    process.exit(1);
  });

